src/classes/BDI_DataImportService.cls

```diff
@@ -1725,68 +1723,141 @@ global with sharing class BDI_DataImportService {
-        }
-
         // need to find all the Campaigns that we might need
-        Set<String> setStrCampaign = new Set<String>();
+        Set<String> uniqueCampaignNames = new Set<String>();
+
         for (DataImport__c dataImport : listDI) {
             // skip di's that already have an error
             if (dataImport.Status__c == BDI_DataImport_API.bdiFailed) {
                 continue;
             }
 
             if (dataImport.Donation_Campaign_Name__c != null) {
-                setStrCampaign.add(dataImport.Donation_Campaign_Name__c);
+                uniqueCampaignNames.add(dataImport.Donation_Campaign_Name__c);
             }
-
         }
 
         // add the existing campaigns to our map
-        for (Campaign cmp : [SELECT Id, Name FROM Campaign WHERE Name IN :setStrCampaign]) {
-            mapStrNameToCampaignId.put(cmp.Name, cmp.Id);
+        List<Campaign> uniqueCampaigns = [
+                SELECT Id, Name
+                FROM Campaign
+                WHERE Name IN :uniqueCampaignNames
+                ORDER BY CreatedDate DESC
+        ];
+
+        for (Campaign cmp : uniqueCampaigns) {
+            campaignNameToCampaignId.put(cmp.Name, cmp.Id);
         }
 
-        // create any new campaigns
-        List<Campaign> listCmpNew = new List<Campaign>();
-        List<DataImport__c> listDINew = new List<DataImport__c>();
-        setStrCampaign.clear();
+        Map<String, Campaign> campaignNameToNewCampaign = new Map<String, Campaign>();
+        Map<String, List<DataImport__c>> newCampaignNameToDataImports = new Map<String, List<DataImport__c>>();
+
+        // create any new campaigns and set status fields
         for (DataImport__c dataImport : listDI) {
             // skip di's that already have an error
             if (dataImport.Status__c == BDI_DataImport_API.bdiFailed) {
                 continue;
             }
 
-            String strName = dataImport.Donation_Campaign_Name__c;
-            Campaign cmp;
-            if (strName != null && setStrCampaign.add(strName)) {
-                if (mapStrNameToCampaignId.get(strName) == null) {
-                    cmp = new Campaign(Name=strName, IsActive=true);
-                    listCmpNew.add(cmp);
-                    listDINew.add(dataImport);
-                }
+            String campaignName = dataImport.Donation_Campaign_Name__c;
+
+            setCampaignImportStatusField(dataImport);
+            setCampaignIdIfMatchedByName(dataImport, campaignName);
+
+            // Create a new Campaign during processing if no match on Name
+            if (!isDryRun && dataImport.DonationCampaignImportStatus__c == System.Label.bdiMatchedNone) {
+
+                createNewCampaignIfUnique(campaignName, campaignNameToNewCampaign);
+                associateCampaignWithDataImport(dataImport, campaignName, newCampaignNameToDataImports);
+
             }
-            if (Test.isRunningTest() && strName != null && strName.startsWith('FailTest')) {
```

Comment:
I looked into this and also a full BDI test without this, and we don't need it. We don't have any 'FailTest's hard coded in there, and there weren't any before. This was something in the previous block, but I think it was just added in here since it's in all the other import blocks.